core/src/main/java/org/elasticsearch/action/admin/indices/upgrade/post/TransportUpgradeAction.java

```diff
@@ -130,8 +151,8 @@ protected ShardUpgradeResponse newShardResponse() {
     @Override
     protected ShardUpgradeResponse shardOperation(ShardUpgradeRequest request) {
         IndexShard indexShard = indicesService.indexServiceSafe(request.shardId().getIndex()).shardSafe(request.shardId().id());
-        org.apache.lucene.util.Version version = indexShard.upgrade(request.upgradeRequest());
-        return new ShardUpgradeResponse(request.shardId(), indexShard.routingEntry().primary(), version);
+        org.apache.lucene.util.Version luceneVersion = indexShard.upgrade(request.upgradeRequest());
+        return new ShardUpgradeResponse(request.shardId(), indexShard.routingEntry().primary(), Version.CURRENT, luceneVersion);
```

Comment:
So, the idea is that luceneVersion track the version of the latest segment. The elasticsearch version tracks the version of metadata, and since we update metadata to match the current version I am using the current version instead of version when index was created. This is the tricky part so I will add some comments here.